extensions/message-ids.md

```diff
@@ -0,0 +1,112 @@
+---
+title: IRCv3 Message IDs
+layout: spec
+work-in-progress: true
+copyrights:
+  -
+    name: "James Wheare"
+    period: 2016
+    email: "james@irccloud.com"
+---
+
+## Notes for implementing work-in-progress version
+
+This is a work-in-progress specification.
+
+Software implementing this work-in-progress specification MUST NOT use the
+unprefixed `msgid` tag name. Instead, implementations SHOULD use
+the `draft/msgid` tag name to be interoperable with other software
+implementing a compatible work-in-progress version.
+
+The final version of the specification will use an unprefixed tag name.
+
+## Introduction
+
+This specification describes a mesage tag indicating a server supplied unique ID for `PRIVMSG` and `NOTICE` messages.
+
+## Motivation
+
+Communication on IRC has historically been limited to a flat structure of sequential messages without unique identifiers shared across clients. As a result, the protocol lacks an ability to establish relationships between messages and other entities. This places limits on client-side enhancements such as reply tracking, message editing, rating or sending other forms of message annotation.
+
+The message ID tag is a way for servers to enable these enhancements.
+
+## Architecture
+
+### Dependencies
+
+The `draft/message-tags` capability MUST be negotiated for servers wishing to use this tag.
+
+### Tags
+
+This specification adds the `draft/msgid` tag, which has one required value.
+
+This tag MAY be sent by a server on `PRIVMSG` and `NOTICE` messages.
+
+#### Tag value
+
+The tag value is chosen by the server and MUST be unique, meaning any other message transmitted on the entire network at any time MUST NOT share the same value.
```

Comment:
That does result in a situation where you can get duplicate messages in a buffer that share an ID (by cycling more than once on a chathistory channel). So I guess clients need to be able to handle that and treat the most recent version as the target.